src/adjoint/adjointDebug.F90

```diff
@@ -0,0 +1,736 @@
+        machGrid = machGrid - h*extraDot(iMachGrid)
+        PinfDim = PinfDim - h*extraDot(iPressure)
+        rhoinfDim = rhoinfDim - h*extraDot(iDensity)
+        tinfdim = tinfdim - h*extraDot(iTemperature)
+        pointref(1) = pointref(1) - h*extraDot(iPointRefX)
+        pointref(2) = pointref(2) - h*extraDot(iPointRefY)
+        pointref(3) = pointref(3) - h*extraDot(iPointRefZ)
+        rgasdim = rgasdim - h*zero
+
+
+
+
+        ! Copy out the residual derivative into the provided dwDot and remove the
+        ! perturbation
+        ii = 0
+        jj = 0
+        do nn=1, nDom
+            do sps=1,nTimeIntervalsSpectral
+            call setPointers_d(nn, 1, sps)
+            do k=1, kl
+                do j=1,jl
+                    do i=1,il
+                        do l=1,3
+                        ii = ii + 1
+                        x(i, j, k, l) = x(i, j, k, l) -  xvDot(ii)* h
+                        end do
+                    end do
+                end do
+            end do
+            do k=2, kl
+                do j=2, jl
+                    do i=2, il
+                        do l=1, nw
+                        jj = jj + 1
+                        w(i, j, k, l) = w(i, j, k, l) -  wDot(jj)*h
+                        dwd(i,j,k,l) = (dwd(i,j,k,l) + dw(i,j,k,l))/h
+                        dwDot(jj) = dwd(i,j,k,l) ! copy values to output
+                        end do
+                    end do
+                end do
+            end do
+
+            end do
+        end do
+
+
+        if(equations == RANSEquations) then
+            call updateXSurf(level)
+        end if
+        bcDataValues = bcDataValues - bcDataValuesdot*h
+
+        
+        &
+        fDot = (fDot + forces)/h
+        funcsDot = (funcsDot + funcValues)/h
+
+    end subroutine computeMatrixFreeProductFwdFD
+
+
+    subroutine printADSeeds(nn, level, sps)
+        ! this routine is used for debugging master_d, and master_b.
+        ! it prints all the AD seeds used
+        ! it is useful to save the output and compare it with a diff tool
+
+
+        use constants
+        use block, only : flowDomsd, flowDoms
+        use blockPointers
+        use inputTimeSpectral
+        use flowVarRefState
+        use inputPhysics
+        use BCPointers_b
+        use communication
+        use oversetData, only : oversetPresent
+        use cgnsGrid, only : cgnsDoms, cgnsDomsd, cgnsNDom
+        use actuatorRegionData, only : nActuatorRegions, actuatorRegionsd
+        implicit none
+
+        ! Input parameters
+        integer(kind=intType) :: nn, level, sps
+
+        ! Working parameters
+        integer(kind=intType) :: mm, i, iDom
+        integer(kind=intType) :: iBoco, iData, iDirichlet
+        write(*,*) 'ADSeeds for block', nn, ' at level ', level, ' at sps ', sps
+        write(*,*) 'd2wall ', minval(flowDomsd(nn, level, sps)%d2wall), &
+                        maxval(flowDomsd(nn, level, sps)%d2wall)
```

Comment:
Besides the max and min, I think the L2 norm gives some idea of the overall match of the data.

Reply:
I agree with @SichengHe here, probably worth doing the L2. 

Also, do you anticipate using these routines in parallel runs? How would these print statements work there? In that case, would we distinguish each value from each proc, or would we report the global values? Printing different stuff from procs w/o paying attention to the order will break the "use it with a diff tool" usecase.

Reply:
L2 added.

I have never used this routine in parallel. 
You'd probably want to redirect the output of each proc to a separate file  (which can be done with a flag).

Reply:
Okay, then as long as you also include these use cases in the documentation update, its fine.